### PR TITLE
[dynamic finetuning] Adding dynamic-finetuning code

### PR DESCRIPTION
**Pull request by kparichay**
Commit 1 : [dyanmic training] Adding dynamic-training code …

Added dynamic-training code with both max and l2norm mode
Verified working with existing examples given the threshold

Commit 2: [dynamic-training] Add dynamic training using derivatives …

Added dynamic training using derivatives where the decision to
apply the gradient is calculated using the derivative received
without calculating the gradient itself.

**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: nntrainer/models/dynamic_training_optimization.h [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+  /**
+   * @brief     Set threshold for optimization
+   */
+  void setThreshold(float threshold_) { threshold = threshold_; };
+
+  /**
+   * @brief     Set the reduce operation for dynamic optimization
+   */
+  void setOp(std::string op) {
```
Comment (zhoonit) [inline nntrainer/models/dynamic_training_optimization.h:45]:
```suggestion
  void setOp(const std::string &op) {
```
How about using const &?

File: nntrainer/models/dynamic_training_optimization.h [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+  /**
+   * @brief     Disable the optimization
+   */
+  void disable() { enabled = false; }
+
+  /**
+   * @brief     Set the mode for optimization
+   */
+  void setMode(std::string mode_) {
```
Comment (zhoonit) [inline nntrainer/models/dynamic_training_optimization.h:67]: same

File: nntrainer/models/dynamic_training_optimization.cpp [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  dist = std::uniform_real_distribution<float>(0.0, 1.0);
+}
+
+/**
+ * @brief     Check if the given weights can skip updating
+ * @note true if should be applied, else false
+ */
+bool DynamicTrainingOptimization::checkIfApply(
+  const std::vector<Weight> &weights, const std::shared_ptr<Var_Grad> input,
+  const std::shared_ptr<Var_Grad> output, const std::shared_ptr<Optimizer> opt,
```
Comment (zhoonit) [inline nntrainer/models/dynamic_training_optimization.cpp:41]: I guess those can be `const std::shared_ptr<T> &` to save locking?

File: nntrainer/models/dynamic_training_optimization.cpp [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+/**
+ * @brief     Check if the given weight can skip updating
+ * @note true if should be applied, else false
+ */
+bool DynamicTrainingOptimization::checkIfApply(
+  const Weight &weight, const std::shared_ptr<Var_Grad> &input,
+  const std::shared_ptr<Var_Grad> &output,
+  const std::shared_ptr<Optimizer> &opt, int iteration) {
+  if (iteration < skip_n_iterations)
```
Comment (lhs8928) [inline nntrainer/models/dynamic_training_optimization.cpp:64]: Already check at line 43

File: nntrainer/models/dynamic_training_optimization.cpp [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    return true;
+
+  if (!weight.getTrainable() || weight.getGradientRef().uninitialized())
```
Comment (lhs8928) [inline nntrainer/models/dynamic_training_optimization.cpp:67]: Isn't it right return false(skip) when trainable is set to false?

File: nntrainer/models/dynamic_training_optimization.cpp [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+  const std::shared_ptr<Var_Grad> output, const std::shared_ptr<Optimizer> opt,
+  int iteration) {
+  if (!enabled || iteration < skip_n_iterations)
+    return true;
+
+  std::vector<bool> apply;
+  apply.reserve(weights.size());
+
+  for (auto const &weight : weights)
+    apply.push_back(checkIfApply(weight, input, output, opt, iteration));
```
Comment (zhoonit) [inline nntrainer/models/dynamic_training_optimization.cpp:50]: instead of apply vector, how about return here when false?

File: nntrainer/models/dynamic_training_optimization.cpp [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+/**
+ * @brief   Check if the update should be applied or skipped
+ * @note true if should be applied, else false
+ */
+bool DynamicTrainingOptimization::checkIfApply(float reduced_ratio,
+                                               float learning_rate) {
+  /**
+   * If the reduced update ratio is higher than 1, then always apply update.
```
Comment (lhs8928) [inline nntrainer/models/dynamic_training_optimization.cpp:108]: According to the code below it seems that reduced update ratio is higher than 1 would it skip the update

File: nntrainer/models/dynamic_training_optimization.h [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+   * @brief     Check if the gradient mode is used for optimization
+   */
+  bool isGradientMode() {
+    if (enabled && calc_ratio_mode == dft_opt_mode_gradient)
+      return true;
+    return false;
+  }
+
+  /**
+   * @brief     Set initial iterations to skip from optimization
```
Comment (lhs8928) [inline nntrainer/models/dynamic_training_optimization.h:96]: One simple question.
Does this variable means that at first n_iteration it want to skip the gradient or apply the gradient?

File: nntrainer/models/dynamic_training_optimization.cpp [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+  return ratio.max_abs();
+}
+
+/**
+ * @brief     Operation to decide if update should be skipped
+ * @note      Calcalate l2 norm of the tensor averaged by its size
+ */
+float DynamicTrainingOptimization::reduceByNorm(Tensor const &ratio) {
+  float l2norm = ratio.l2norm();
+  return (l2norm * l2norm) / ratio.length();
```
Comment (lhs8928) [inline nntrainer/models/dynamic_training_optimization.cpp:132]: Is there any reason to square l2norm?

File: nntrainer/models/dynamic_training_optimization.h [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    else if (mode_ == dft_opt_mode_gradient)
+      calc_ratio_op = ratioUsingGradient;
+    else
+      enabled = false;
+  }
+
+  /**
+   * @brief     Check if the derivative mode is used for optimization
+   */
+  bool isDerivativeMode() {
```
Comment (lhs8928) [inline nntrainer/models/dynamic_training_optimization.h:80]: What is the derivative mode and when it is used?

File: nntrainer/models/dynamic_training_optimization.h (hunk 1 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.h:45]: done

File: nntrainer/models/dynamic_training_optimization.h (hunk 2 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.h:67]: done

File: nntrainer/models/dynamic_training_optimization.cpp (hunk 6 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.cpp:50]: currently, all the values are accumulated with `logical_and`, but this is not always true.
In the final version (which will be next PR), weights and biases are separated.
So, this will be completely refactored there.

File: nntrainer/models/dynamic_training_optimization.cpp (hunk 4 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.cpp:64]: This function is also public and can be called from outside.
Hence the second check

File: nntrainer/models/neuralnet.cpp [hunk 11]
```diff
@@ -346,25 +378,20 @@ void NeuralNetwork::backwarding(int iteration) {
...
    * The last trainable layer need not calculate the derivatives
-   * Do not change this order:
-   * 1. calcGradient
-   * 2. calcDerivative
-   * 3. applyGradient
    */
-  last_layer->calcGradient();
 #ifdef ENABLE_TEST
-  last_layer->calcDerivative();
+  backwarding((iter_end - 1)->layer, iteration, true);
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/models/neuralnet.cpp:391]: udpated.

File: nntrainer/models/dynamic_training_optimization.h (hunk 10 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.h:80]: added in the description.

File: nntrainer/models/dynamic_training_optimization.cpp (hunk 9 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.cpp:132]:
```python3
l2norm =  sqrt(sum([(x^2) for x in data]))   # python based pseudo code
l2norm / length = sqrt(... / (length ^ 2)) 
```

As you can see, simply l2norm by length actually divides by `length^2`.

File: nntrainer/models/dynamic_training_optimization.h (hunk 8 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.h:96]: for `first_n_iteration`, dynamic training optimization will be skipped, and gradients will always be applied.
Updated description to be clearer.

File: nntrainer/models/dynamic_training_optimization.cpp (hunk 7 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.cpp:108]: This was inverted. Updated the return values.

File: nntrainer/models/dynamic_training_optimization.cpp (hunk 5 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.cpp:67]: If trainable is set to false, dynamic training optimization cannot make the decision to skip the update or not as the gradient/derivative might not be available.
So, the decision making here is to go with default behavior as if there was no dynamic optimization, which is to return true in this case.

Please let me know if this is counter-intuitive.

File: nntrainer/models/dynamic_training_optimization.cpp (hunk 9 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.cpp:132]: I updated to sqrt of the whole thing.

File: nntrainer/models/dynamic_training_optimization.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.cpp:41]: ohh, i missed in this function. Updated.

File: nntrainer/models/dynamic_training_optimization.cpp (hunk 6 above)
Comment (zhoonit) [reply nntrainer/models/dynamic_training_optimization.cpp:50]: Noted.

File: nntrainer/models/dynamic_training_optimization.h (hunk 10 above)
Comment (lhs8928) [reply nntrainer/models/dynamic_training_optimization.h:103]: Got it.
But it is not used. Will use this in next PR?

File: nntrainer/models/dynamic_training_optimization.cpp (hunk 5 above)
Comment (lhs8928) [reply nntrainer/models/dynamic_training_optimization.cpp:67]: Understood.

File: nntrainer/models/dynamic_training_optimization.h (hunk 10 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.h:103]: aah yes.

File: nntrainer/models/dynamic_training_optimization.h [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+public:
+  /**
+   * @brief     Constructor of DynamicFineTuning Optimization
+   */
+  DynamicTrainingOptimization(int threshold_ = 1, int skip_n_iter = 1);
+
+  /**
+   * @brief     Set threshold for optimization
+   */
+  void setThreshold(float threshold_) { threshold = threshold_; };
```
Comment (jijoongmoon) [inline nntrainer/models/dynamic_training_optimization.h:61]: need to add check if the threshold is greater than 1?

File: nntrainer/models/dynamic_training_optimization.h [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+
+  /**
+   * @brief     Set the reduce operation for dynamic optimization
+   */
+  void setOp(const std::string &op) {
+    if (op == dft_opt_max)
```
Comment (jijoongmoon) [inline nntrainer/models/dynamic_training_optimization.h:67]: may be, provide some notification? if the op is not proper?

File: nntrainer/models/dynamic_training_optimization.h [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+   * @brief     Set the mode for optimization
+   */
+  void setMode(const std::string &mode_) {
+    calc_ratio_mode = mode_;
+    if (mode_ == dft_opt_mode_derivative)
+      calc_ratio_op = ratioUsingDerivative;
+    else if (mode_ == dft_opt_mode_gradient)
+      calc_ratio_op = ratioUsingGradient;
+    else
+      enabled = false;
```
Comment (jijoongmoon) [inline nntrainer/models/dynamic_training_optimization.h:95]: give proper notification?

File: nntrainer/models/dynamic_training_optimization.cpp [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    return true;
+
+  float reduced_ratio = calc_ratio_op(weight, input, output, reduce_op);
+
+  return checkIfApply(reduced_ratio, (float)opt->getLearningRate(iteration));
+}
+
+/**
+ * @brief   Calculate the ratio of update to the weight using derivative
+ */
```
Comment (jijoongmoon) [inline nntrainer/models/dynamic_training_optimization.cpp:77]: better to add comments about input is actual input and output is output about forwarding direction.

File: nntrainer/models/dynamic_training_optimization.h (hunk 12 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.h:61]: Threshold is used a divider against the reduced value. 
So threshold can be anything but zero.

File: nntrainer/models/dynamic_training_optimization.cpp [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+float DynamicTrainingOptimization::ratioUsingDerivative(
+  const Weight &weight, const std::shared_ptr<Var_Grad> &input,
+  const std::shared_ptr<Var_Grad> &output,
+  std::function<float(Tensor const &)> reduce_op) {
+  float reduced_derivative = reduce_op(output->getGradientRef());
```
Comment (jijoongmoon) [inline nntrainer/models/dynamic_training_optimization.cpp:82]: is it reduced already in batch direction?

File: nntrainer/models/dynamic_training_optimization.cpp [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+ */
+bool DynamicTrainingOptimization::checkIfApply(float reduced_ratio,
+                                               float learning_rate) {
+  /**
+   * If the reduced update ratio is higher than 1, then always apply update.
+   * If the reduced update raito is less than 1, then apply it with
+   * probability = update ratio
+   */
+  if (dist(rng) < reduced_ratio * learning_rate / threshold)
+    return true;
```
Comment (jijoongmoon) [inline nntrainer/models/dynamic_training_optimization.cpp:113]: need to check if threshold is greater than 0.

File: nntrainer/models/dynamic_training_optimization.h (hunk 13 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.h:67]: Will add throw.

File: nntrainer/models/dynamic_training_optimization.h (hunk 14 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.h:95]: will add throw.

File: nntrainer/models/dynamic_training_optimization.cpp (hunk 16 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.cpp:82]: This is the old method. The final method using approximate gradient over batch will be added in next PR.

File: nntrainer/models/dynamic_training_optimization.cpp (hunk 17 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.cpp:113]: Added while setting the threshold.

File: nntrainer/models/dynamic_training_optimization.cpp (hunk 15 above)
Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.cpp:77]: added.

File: nntrainer/models/dynamic_training_optimization.h (hunk 12 above)
Comment (jijoongmoon) [reply nntrainer/models/dynamic_training_optimization.h]: I think at least we have to give them a notification what threshold which greater than 1.0 means. if the threshold is greater than 1, then it means only layer which approximation of gradient is greater than weight is updated. IMO, most of layer does not updated in normal cases.

Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.h]: Then that notification should be give for even smaller thresholds.
Threshold of 1e-3 for resnet50 led to 75% skip of gradient while 1e-3 for VGG16 led to over 90% skip.
So, I am not sure what notification can be given for threshold over 1, as there is no direct meaning to the threshold values by itself.

Comment (jijoongmoon) [reply nntrainer/models/dynamic_training_optimization.h]: does it means most layers are going to update if threshold over 1 is used.  Am I right? If it is, then it is kind of the opposite of user's intention.

Comment (kparichay) [reply nntrainer/models/dynamic_training_optimization.h]: So, the value of threshold cannot be directly compared to 1. Its range is a lot dependent on model, data and training configuration.

File: nntrainer/tensor/var_grad.h [hunk 18]
```diff
@@ -211,6 +211,20 @@ class Var_Grad {
    */
   Tensor &getGradientRef() { return *grad.get(); }
 
+  /**
+   * @brief Get the variable tensor (by reference)
+   *
+   * @return Tensor Variable tensor
+   */
+  Tensor &getVariableRef() const { return *var.get(); }
```
Comment (zhoonit) [inline nntrainer/tensor/var_grad.h:219]: return const reference? (and also for :226)

Comment (kparichay) [reply nntrainer/tensor/var_grad.h:236]: In many cases, we need to modify the tensor or call in_place operations on this.
That won't be possible with const.

Comment (zhoonit) [reply nntrainer/tensor/var_grad.h:236]: If modifying tensor take places, shouldn't it be `var_grad` rather than `const var_grad` in the first place?

Comment (kparichay) [reply nntrainer/tensor/var_grad.h:236]: done.

Comment (zhoonit) [reply nntrainer/tensor/var_grad.h]: Glad that it didn't.

**Review by zhoonit — APPROVED** [review]
Great work man 👍

File: nntrainer/models/dynamic_training_optimization.h (hunk 12 above)
Comment (jijoongmoon) [reply nntrainer/models/dynamic_training_optimization.h]: > Its range is a lot dependent on model, data and training configuration.

Well, if you say so.. let's see. leave as it is.

**Review by jijoongmoon — APPROVED** [review]
LGTM
